File: JavaWorld/src/com/java/common/MergeSort.java
package com.java.common;

import java.util.Arrays;

class MergeSort {

    public static void main(String... args) {
        int[] arr = {5, 10, 15, 0, -9, 20, 99, 17, 14};

        System.out.println(Arrays.toString(merge_sort(arr)));
    }

    private static int[] merge_sort(int[] arr) {
        if(arr.length <= 1) return arr;
        int mid = arr.length / 2;

        int[] first = new int[mid];
        int[] second = new int[arr.length - mid];
        for(int i = 0; i < arr.length; i++) {
            if(i < mid) {
                first[i] = arr[i];
            } else {
                second[i-mid] = arr[i];
            }
        }

        return merge(merge_sort(first), merge_sort(second));
    }

    private static int[] merge(int[] first, int[] second) {
        System.out.println("Merge: " + Arrays.toString(first) + " & " + Arrays.toString(second));
        int[] arr = new int[first.length + second.length];
        int i=0, j=0, idx=0;

        while (i<first.length || j<second.length) {
            if(i == first.length) {
                arr[idx++] = second[j++];
                continue;
            }

            if(j == second.length) {
                arr[idx++] = first[i++];
                continue;
            }

            arr[idx++] = first[i] > second[j] ? second[j++] : first[i++];
        }

        return arr;
    }
}

File: JavaWorld/src/com/kotlin/MergeSort.kt
package com.kotlin

fun main(args: Array<String>) {
    val arr = arrayOf(10,5,19,7,0,15,29,23,10)

    println(mergeSort(arr).contentToString())
}

fun mergeSort(arr: Array<Int>): Array<Int> {
    if (arr.size <= 1) return arr
    val mid = arr.size / 2
    val first = arr.filterIndexed{ idx, _ -> idx < mid }.toTypedArray()
    val second = arr.filterIndexed{ idx, _ -> idx >= mid }.toTypedArray()

    return merge(mergeSort(first), mergeSort(second))
}

fun merge(first: Array<Int>, second: Array<Int>): Array<Int> {
    println("Merge ${first.contentToString()} + ${second.contentToString()}")
    var i = 0
    var j = 0
    var arr = arrayOf<Int>()

    while(i<first.size || j<second.size) {
        if(i == first.size) {
            arr += second[j++]
            continue
        }
        if(j == second.size) {
            arr += first[i++]
            continue
        }

        arr += if(first[i] > second[j]) second[j++] else first[i++]
    }
    return arr
}